### PR TITLE
Add ability to group notes by tags

### PR DESCRIPTION
**Pull request by t-eliks**

### CODE REVIEW

**Comment by t-eliks** [comment]
[Add grouping for notes](https://trello.com/c/AOIvNfOa/30-add-grouping-for-notes)

File: ToNote/ViewModels/MainViewModel.cs [hunk 1]
```diff
@@ -142,5 +155,77 @@ public ICommand OpenAboutPageCommand
...
+                if (_SelectedTags != value)
+                {
+                    _SelectedTags = value;
+
+                    RaisePropertyChanged(nameof(SelectedTags));
+                }
+            }
+        }
+
+        public IEnumerable AllTags { get => Notes.SelectMany(x => x.Tags).Distinct(); }
```
Comment (Kodnot) [inline ToNote/ViewModels/MainViewModel.cs]: Could be written as `public IEnumerable AllTags => Notes.SelectMany(x => x.Tags).Distinct();`

File: ToNote/ViewModels/MainViewModel.cs [hunk 2]
```diff
@@ -142,5 +155,77 @@ public ICommand OpenAboutPageCommand
...
+                {
+                    if (SelectedTags.Contains(tag))
+                        SelectedTags.Remove(tag);
+                    else
+                        SelectedTags.Add(tag);
+                }));
+            }
+        }
+
+        public ICollectionView FilteredNotes
```
Comment (Kodnot) [inline ToNote/ViewModels/MainViewModel.cs]: `public ICollectionView FilteredNotes => CollectionViewSource.GetDefaultView(Notes);`

File: ToNote/Views/MainView.xaml [hunk 3]
```diff
@@ -50,13 +50,35 @@
...
         <ScrollViewer HorizontalScrollBarVisibility="Auto" VerticalScrollBarVisibility="Auto" Grid.Row="1" Style="{StaticResource RightScrollViewer}">
-            <ItemsControl Margin="25" ItemsSource="{Binding Notes}">
-                <ItemsControl.ItemsPanel>
-                    <ItemsPanelTemplate>
-                        <StackPanel IsItemsHost="True"/>
-                    </ItemsPanelTemplate>
-                </ItemsControl.ItemsPanel>
-            </ItemsControl>
+            <DockPanel>
+                <StackPanel DockPanel.Dock="Top" Orientation="Horizontal">
```
Comment (Kodnot) [inline ToNote/Views/MainView.xaml]: Maybe add a margin, the dropdown list is too close to the left border of the app, it looks a bit ugly

File: ToNote/Views/MainView.xaml [hunk 4]
```diff
@@ -50,13 +50,35 @@
...
+                    <Button Content="Select tags" Command="{Binding ToggleGroupingPanelCommand}"/>
+                    <Popup IsOpen="{Binding IsGroupingPanelOpen, Mode=TwoWay}" StaysOpen="False">
```
Comment (Kodnot) [inline ToNote/Views/MainView.xaml]: This approach feels a bit hacky to me. Why didn't you use a ComboBox or something? What's the reasoning behind implementing the dropdown via a popup?

**Review by Kodnot — CHANGES_REQUESTED** [review]

**Comment by Kodnot** [comment]
You broke dragging notes around, the program now crashes on drag. When you fix it, check if it would be possible to unit test the issue, as this part of our code breaks often and it would be nice to have some tests to prevent more regressions.

File: ToNote/Views/MainView.xaml (hunk 4 above)
Comment (Kodnot) [reply ToNote/Views/MainView.xaml]: The reasoning was provided on slack, this is more flexible. Still feels a bit hacky to me, the implementation could be smoother, but it'll do.

File: ToNote/Views/MainView.xaml (hunk 3 above)
Comment (t-eliks) [reply ToNote/Views/MainView.xaml]: Wasn't concerned with the look part of it, left it up for whoever styles it to come up with the best placement for it

File: ToNote/ViewModels/MainViewModel.cs [hunk 5]
```diff
@@ -18,6 +21,14 @@ public MainViewModel()
             {
                 using (var reader = new StreamReader(file))
                     Notes.Add(JsonConvert.DeserializeObject<Note>(reader.ReadToEnd()));
+
+                SelectedTags.CollectionChanged += (s, e) =>
+                {
+                    if (SelectedTags.Any())
```
Comment (Kodnot) [inline ToNote/ViewModels/MainViewModel.cs]: You don't need this check. col.All(pred) will return true for an empty collection, so you can leave just `FilteredNotes.Filter = x => SelectedTags.All(t => ((Note)x).Tags.Contains(t));` 
There shouldn't be any performance concerns, so no reason not to prefer a more concise version.

File: ToNoteTests/Core/VisualTreeService.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                if (childType == null)
+                {
+                    // recursively drill down the tree
+                    foundChild = FindChild<T>(child, predicate);
+                    // If the child is found, break so we do not overwrite the found child. 
+                    if (foundChild != null) break;
+                }
+                else if (predicate == null)
+                {
+                    {
```
Comment (Kodnot) [inline ToNoteTests/Core/VisualTreeService.cs]: Why the double {}?

**Review by Kodnot — APPROVED** [review]
Some minor changes would be nice, but otherwise great job

**Pull request merged**